haystack/pipelines/config.py

```diff
@@ -129,6 +123,17 @@ def build_component_dependency_graph(
                 referenced_components.append(param_value)
         for referenced_component in referenced_components:
             graph.add_edge(referenced_component, component_name)
+    for node in pipeline_definition["nodes"]:
+        node_name = node["name"]
+        graph.add_node(node_name)
+        for input in node["inputs"]:
+            if input in component_definitions:
+                # Special case for (actually permitted) cyclic dependencies between two components:
+                # e.g. DensePassageRetriever depends on ElasticsearchDocumentStore.
+                # In indexing pipelines ElasticsearchDocumentStore depends on DensePassageRetriever's output.
+                # But this second dependency is looser, so we neglect it.
+                if not graph.has_edge(node_name, input):
```

Comment:
Just a check whether I understand this correctly: As an example, in an indexing pipeline `ElasticsearchDocumentStore` could be the `node` and `DensePassageRetriever` could be the `input`. Then we check here whether there is an edge in the graph from `ElasticsearchDocumentStore` to `DensePassageRetriever` and if that's not the case then we add an edge with the opposite direction, correct?

Reply:
Yes, correct.

Reply:
Just a curiosity: why we need to know what are the runtime dependencies between the nodes at this stage? I thought this function is used in the codegen code to create the nodes in a working order (so nodes that are required as input will be initialized first). The fact that later on in the pipeline a node will get output from another should not matter... right?

Reply:
Yes, right. I chose to include those runtime dependencies to make the order of init calls more readable. With it it'll be in pipeline execution order (like a human would normally instantiate it), without it it would be in arbitrary order and less comprehensive for humans.